File: go/2024/solutions/day05.go
package solutions

import (
	"fmt"
	"slices"
	"strconv"
	"strings"
)

type Day05 struct{}
type input05 struct {
	raw    string
	rules  []rule
	prints [][]int
}
type rule struct{ a, b int }

func (d Day05) Parse(raw string) input05 {
	parts := strings.Split(raw, "\n\n")
	rawRules := strings.Split(parts[0], "\n")
	rawPrints := strings.Split(parts[1], "\n")

	rules := make([]rule, len(rawRules))
	for i, r := range rawRules {
		parts := strings.Split(r, "|")
		a, _ := strconv.Atoi(parts[0])
		b, _ := strconv.Atoi(parts[1])
		rules[i] = rule{a, b}
	}
	prints := make([][]int, len(rawPrints))
	for i, p := range rawPrints {
		parts := strings.Split(p, ",")
		prints[i] = make([]int, len(parts))
		for j, sn := range parts {
			n, _ := strconv.Atoi(sn)
			prints[i][j] = n
		}
	}
	return input05{raw, rules, prints}
}

func (d Day05) Part1(raw string) {
	fmt.Println("Day 5 Part 1")
	input := d.Parse(raw)
	sum := 0

printLoop:
	for _, print := range input.prints {
		for _, rule := range input.rules {
			if posB := slices.IndexFunc(print, match(rule.b)); posB >= 0 {
				if posA := slices.IndexFunc(print, match(rule.a)); posA >= 0 && posB < posA {
					continue printLoop
				}
			}
		}
		// all rules pass
		sum += print[int(len(print)/2)]
	}

	fmt.Println("Solution:", sum)
}

func (d Day05) Part2(raw string) {
	fmt.Println("Day 5 Part 2")
	input := d.Parse(raw)
	sum := 0

	for _, print := range input.prints {
		include := false
		check := true
		for check {
			check = false
			for _, rule := range input.rules {
				if posB := slices.IndexFunc(print, match(rule.b)); posB >= 0 {
					if posA := slices.IndexFunc(print, match(rule.a)); posA >= 0 && posB < posA {
						include = true
						check = true
						print[posA], print[posB] = print[posB], print[posA]
					}
				}
			}
		}
		if include {
			sum += print[int(len(print)/2)]
		}
	}

	fmt.Println("Solution:", sum)
}

func match(n int) func(int) bool {
	return func(element int) bool {
		return element == n
	}
}

File: rust/2024/src/solutions/day_05.rs
use itertools::Itertools;

use super::day_trait::Day;

#[derive(Debug)]
pub struct Input {
    lines: Vec<String>,
    rules: Vec<(usize, usize)>,
    books: Vec<Vec<usize>>,
}

fn parse_input(raw: &str) -> Input {
    let lines: Vec<String> = raw.split("\n").map(|s| s.to_owned()).collect();
    let (raw_rules, raw_books) = raw.split_once("\n\n").unwrap();
    let rules = raw_rules
        .split("\n")
        .map(|r| {
            r.split_once("|")
                .map(|(a, b)| (a.parse().unwrap(), b.parse().unwrap()))
                .unwrap()
        })
        .collect_vec();
    let books = raw_books
        .split("\n")
        .map(|b| b.split(",").map(|n| n.parse().unwrap()).collect_vec())
        .collect_vec();

    Input {
        lines,
        rules,
        books,
    }
}

pub struct Day05;
impl Day for Day05 {
    fn day(&self) -> u8 {
        05
    }

    fn part_1(&self, raw: &str) {
        println!("Day {} part 1", self.day());
        let input: Input = parse_input(raw);

        let sum: usize = input
            .books
            .iter()
            .filter(|&book| {
                input.rules.iter().all(|rule| {
                    if let (Some(p1), Some(p2)) = (
                        book.iter().position(|&n| n == rule.0),
                        book.iter().position(|&n| n == rule.1),
                    ) {
                        p1 < p2
                    } else {
                        true
                    }
                })
            })
            .map(|book| book[book.len() / 2])
            .sum();

        println!("Answer is {}", sum);
    }

    fn part_2(&self, raw: &str) {
        println!("Day {} part 2", self.day());
        let input: Input = parse_input(raw);

        let sum: usize = input
            .books
            .iter()
            .filter_map(|book| {
                let mut corrected_book = book.clone();
                let mut corrected = false;
                let mut modified = true;
                while modified {
                    modified = false;
                    for rule in &input.rules {
                        if let (Some(p1), Some(p2)) = (
                            corrected_book.iter().position(|&n| n == rule.0),
                            corrected_book.iter().position(|&n| n == rule.1),
                        ) {
                            if p1 > p2 {
                                corrected = true;
                                modified = true;
                                (corrected_book[p1], corrected_book[p2]) =
                                    (corrected_book[p2], corrected_book[p1])
                            }
                        }
                    }
                }
                match corrected {
                    true => Some(corrected_book[corrected_book.len() / 2]),
                    false => None,
                }
            })
            .sum();

        println!("Answer is {}", sum);
    }
}
